frontend/src/components/NewRunParameters.tsx

```diff
@@ -39,23 +61,145 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
         <div>{titleMessage}</div>
         {!!initialParams.length && (
           <div>
-            {initialParams.map((param, i) => (
-              <TextField
-                id={`newRunPipelineParam${i}`}
-                key={i}
-                variant='outlined'
-                label={param.name}
-                value={param.value || ''}
-                onChange={ev => handleParamChange(i, ev.target.value || '')}
-                style={{ maxWidth: 600 }}
-                className={commonCss.textField}
-              />
-            ))}
+            {initialParams.map((param, i) => {
+              return (
+                <EnhancedTextField
+                  key={i}
+                  param={param}
+                  index={i}
+                  handleParamChange={handleParamChange}
+                />
+              );
+            })}
           </div>
         )}
       </div>
     );
   }
 }
 
+interface EnhancedTextFieldProps {
+  param: ApiParameter;
+  index: number;
+  handleParamChange: (index: number, value: string) => void;
+}
+
+interface EnhancedTextFieldState {
+  isEditorOpen: boolean;
+  isInJsonForm: boolean;
+  isJsonField: boolean;
+}
+
+class EnhancedTextField extends React.Component<EnhancedTextFieldProps, EnhancedTextFieldState> {
+  public static getDerivedStateFromProps(
+    nextProps: EnhancedTextFieldProps,
+    prevState: EnhancedTextFieldState,
+  ): EnhancedTextFieldState {
+    let isJson = true;
+    try {
+      const displayValue = JSON.parse(nextProps.param.value || '');
+      // Nulls, booleans, strings, and numbers can all be parsed as JSON, but we don't care
+      // about rendering. Note that `typeOf null` returns 'object'
+      if (displayValue === null || typeof displayValue !== 'object') {
+        throw new Error('Parsed JSON was neither an array nor an object. Using default renderer');
+      }
+    } catch (err) {
+      isJson = false;
+    }
+    return {
+      isEditorOpen: prevState.isEditorOpen,
+      isInJsonForm: isJson,
+      isJsonField: prevState.isJsonField || isJson,
+    };
+  }
+
+  constructor(props: any) {
+    super(props);
+
+    this.state = {
+      isEditorOpen: false,
+      isInJsonForm: false,
+      isJsonField: false,
+    };
+  }
+
+  public render(): JSX.Element | null {
+    const { param, index, handleParamChange } = this.props;
+
+    if (this.state.isJsonField) {
+      return (
+        <>
+          <TextField
+            id={`newRunPipelineParam${index}`}
+            disabled={this.state.isEditorOpen}
+            variant='outlined'
+            label={param.name}
+            value={param.value || ''}
+            onChange={ev => handleParamChange(index, ev.target.value || '')}
```

Comment:
nit: these calls to `handleParamChange` with an index is leaking implementation details. EnhancedTextField should be a complete component that can be used by itself (although it's fine we put them in the same file for now).

Recommend renaming the prop to `onChange` and accepts just the value.